File: 04.Seminar/lesson4/src/main/java/ru/geekbrains/lesson4/HashMap.java
package ru.geekbrains.lesson4;

public class HashMap<K, V> {

    private static final int INIT_BUCKET_COUNT = 16;
    private static final double LOAD_FACTOR = 0.5; //50%
    private int size;

    private Bucket[] buckets;

    class Entity {
        K key;
        V value;
    }

    class Bucket<K, V> {

        Node head;

        class Node {
            Node next;

            Entity nodeValue;
        }

        public V add(Entity entity) {
            Node node = new Node();
            node.nodeValue = entity;

            if (head == null) {
                head = node;
                return null;
            }
            Node currentNode = head;
            while (true) {
                if (currentNode.nodeValue.key.equals(entity.key)) {
                    V buf = (V) currentNode.nodeValue.value;
                    currentNode.nodeValue.value = entity.value;
                    return buf;
                }
                if (currentNode.next != null) {
                    currentNode = currentNode.next;
                } else {
                    currentNode.next = node;
                    return null;
                }
            }
        }

        public V get(K key) {
            Node node = head;
            while (node != null) {
                if (node.nodeValue.key.equals(key)) {
                    return (V) node.nodeValue.value;
                }
                node = node.next;
            }
            return null;
        }

        public V remove(K key) {
            if (head == null)
                return null;
            if (head.nodeValue.key.equals(key)) {
                V buf = (V) head.nodeValue.value;
                head = head.next;
                return buf;
            } else {
                Node node = head;
                while (node.next != null) {
                    if (node.next.nodeValue.key.equals(key)) {
                        V buf = (V) node.next.nodeValue.value;
                        node.next = node.next.next;
                        return buf;
                    }
                    node = node.next;
                }
                return null;
            }
        }


    }

    private int calculateBucketIndex(K key) {
        return Math.abs(key.hashCode()) % buckets.length;
    }

    private void recalculate(){
        size = 0;
        Bucket<K, V>[] old = buckets;
        buckets = new Bucket[old.length * 2];
        for (int i = 0; i < old.length; i++){
            Bucket<K, V> bucket = old[i];
            if (bucket != null){
                Bucket.Node node = bucket.head;
                while (node != null){
                    put((K)node.nodeValue.key, (V)node.nodeValue.value);
                    node = node.next;
                }
            }
            old[i] = null;
        }
    }

    public V put(K key, V value) {

        if (buckets.length * LOAD_FACTOR <= size)
            recalculate();

        int index = calculateBucketIndex(key);
        Bucket bucket = buckets[index];
        if (bucket == null) {
            bucket = new Bucket();
            buckets[index] = bucket;
        }
        Entity entity = new Entity();
        entity.key = key;
        entity.value = value;
        V ret = (V)bucket.add(entity);
        if (ret == null){
            size++;
        }
        return ret;
    }

    public V get(K key) {
        int index = calculateBucketIndex(key);
        Bucket bucket = buckets[index];
        if (bucket == null)
            return null;
        return (V) bucket.get(key);
    }

    public V remove(K key) {
        int index = calculateBucketIndex(key);
        Bucket bucket = buckets[index];
        if (bucket == null)
            return null;
        V ret = (V) bucket.remove(key);
        if (ret != null){
            size--;
        }
        return ret;
    }

    public HashMap() {
        this(INIT_BUCKET_COUNT);
        //buckets = new Bucket[INIT_BUCKET_COUNT];
    }

    public HashMap(int initCount) {
        buckets = new Bucket[initCount];
    }


}

File: 04.Tasks/Task4/HashMap.cs
using System.Text;

namespace Task4
{
    public class HashMap<K, V>
    {
        private static readonly int INIT_BUCKET_COUNT = 16;
        private static readonly double LOAD_FACTOR = 1; 
        private int size;
        public Bucket<K,V>[] buckets;
        public class Entity
        {
            public K? key;
            public V? value;
        }
        public class Bucket<I, J>
        {
            public Node? head;
            public class Node
            {
                public Node? next;

                public Entity? nodeValue;
            }
            public V Add(Entity entity)
            {
                Node node = new Node();
                node.nodeValue = entity;

                if (head == null)
                {
                    head = node;
                    return default(V);
                }
                Node currentNode = head;
                while (true)
                {
                    if (currentNode.nodeValue.key.Equals(entity.key))
                    {
                        V buf = currentNode.nodeValue.value;
                        currentNode.nodeValue.value = entity.value;
                        return buf;
                    }
                    if (currentNode.next != null)
                    {
                        currentNode = currentNode.next;
                    }
                    else
                    {
                        currentNode.next = node;
                        return default(V);
                    }
                }
            }
            public V Get(K key)
            {
                Node node = head;
                while (node != null)
                {
                    if (node.nodeValue.key.Equals(key))
                    {
                        return node.nodeValue.value;
                    }
                    node = node.next;
                }
                return default(V);
            }
            public V Remove(K key)
            {
                if (head == null)
                    return default(V);
                if (head.nodeValue.key.Equals(key))
                {
                    V buf = head.nodeValue.value;
                    head = head.next;
                    return buf;
                }
                else
                {
                    Node node = head;
                    while (node.next != null)
                    {
                        if (node.next.nodeValue.key.Equals(key))
                        {
                            V buf = node.next.nodeValue.value;
                            node.next = node.next.next;
                            return buf;
                        }
                        node = node.next;
                    }
                    return default(V);
                }
            }

        }
        public int CalculateBucketIndex(K key)
        {
            return Math.Abs(key.GetHashCode()) % buckets.Length;
        }
        public void Recalculate()
        {
            size = 0;
            Bucket<K, V>[] old = buckets;
            buckets = new Bucket<K, V>[old.Length * 2];
            for (int i = 0; i < old.Length; i++)
            {
                Bucket<K, V> bucket = old[i];
                if (bucket != null)
                {
                    Bucket<K,V>.Node node = bucket.head;
                    while (node != null)
                    {
                        Put(node.nodeValue.key, node.nodeValue.value);
                        node = node.next;
                    }
                }
                old[i] = null;
            }
        }
        public V Put(K key, V value)
        {
            if (buckets.Length * LOAD_FACTOR <= size)
                Recalculate();
            int index = CalculateBucketIndex(key);
            Bucket<K, V> bucket = buckets[index];
            if (bucket == null)
            {
                bucket = new Bucket<K, V>();
                buckets[index] = bucket;
            }
            Entity entity = new Entity();
            entity.key = key;
            entity.value = value;
            V ret = bucket.Add(entity);
            if (ret == null)
            {
                size++;
            }
            return ret;
        }
        public V Get(K key)
        {
            int index = CalculateBucketIndex(key);
            Bucket<K, V> bucket = buckets[index];
            if (bucket == null)
                return default(V);
            return bucket.Get(key);
        }
        public V Remove(K key)
        {
            int index = CalculateBucketIndex(key);
            Bucket<K, V> bucket = buckets[index];
            if (bucket == null)
                return default(V);
            V ret = bucket.Remove(key);
            if (ret != null)
            {
                size--;
            }
            return ret;
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            bool first = true;
            foreach (var bucket in buckets)
            {
                if (bucket != null)
                {
                    var node = bucket.head;
                    while (node != null)
                    {
                        if (first)
                        {
                            first = false;
                        }
                        else
                        {
                            sb.Append("\n");
                        }
                        sb.Append(node.nodeValue.key);
                        sb.Append(", ");
                        sb.Append(node.nodeValue.value);
                        node = node.next;
                    }
                }
            }
            return sb.ToString();
        }

        public string toPrint()
        {
            StringBuilder sb = new StringBuilder();
            bool first = true;
            for (int i = 0; i < buckets.Length; i++)
            {
                var bucket = buckets[i];
                if (bucket != null)
                {
                    var node = bucket.head;
                    int nodeNumber = 0;
                    while (node != null)
                    {
                        if (first)
                        {
                            first = false;
                        }
                        else
                        {
                            sb.Append("\n");
                        }
                        sb.Append("Bucket: ");
                        sb.Append(i);
                        sb.Append(", Node: ");
                        sb.Append(nodeNumber); 
                        sb.Append(", Key: ");
                        sb.Append(node.nodeValue.key);
                        sb.Append(", Value: ");
                        sb.Append(node.nodeValue.value);
                        node = node.next;
                        nodeNumber++; 
                    }
                }
            }
            return sb.ToString();
        }


        public HashMap() : this(INIT_BUCKET_COUNT)
        {

        }
        public HashMap(int initCount)
        {
            buckets = new Bucket<K, V>[initCount];
        }
    }
}
